src/SIL.XForge.Scripture/Services/SFProjectService.cs

```diff
@@ -571,71 +602,86 @@ public async Task<IReadOnlyList<InviteeStatus>> InvitedUsersAsync(string curUser
+
+        if (projectSecretShareKey.RecipientUserId != null)
+        {
+            if (projectSecretShareKey.RecipientUserId == curUserId)
+            {
+                return projectId;
+            }
+            throw new ForbiddenException();
+        }
+
         IDocument<SFProject> projectDoc = await GetProjectDocAsync(projectId, conn);
         SFProject project = projectDoc.Data;
         if (project.UserRoles.ContainsKey(curUserId))
-            return;
+            return projectId;
 
         IDocument<User> userDoc = await conn.FetchAsync<User>(curUserId);
         // Attempt to get the role for the user from the Paratext registry
         Attempt<string> attempt = await TryGetProjectRoleAsync(project, curUserId);
         if (attempt.TryResult(out string projectRole))
         {
-            // Add the user and remove the specific user share key if it exists.
-            await AddUserToProjectAsync(conn, projectDoc, userDoc, projectRole, true);
-            return;
+            await AddUserToProjectAsync(conn, projectDoc, userDoc, projectRole, projectSecretShareKey.Key);
+            return projectId;
         }
 
-        // Get the project role that is specified in the sharekey
-        Attempt<SFProjectSecret> psAttempt = await ProjectSecrets.TryGetAsync(projectId);
-        if (psAttempt.TryResult(out SFProjectSecret ps))
-            projectRole = ps.ShareKeys.SingleOrDefault(sk => sk.Key == shareKey)?.ProjectRole;
-        // The share key was invalid
-        if (projectRole == null)
+        // Get the project role that is specified in the shareKey
+        if (projectSecretShareKey.ProjectRole == null)
             throw new ForbiddenException();
 
-        string[] availableRoles = new Dictionary<string, bool>
+        if (projectSecretShareKey.ShareLinkType == ShareLinkType.Anyone)
         {
+            string[] availableRoles = new Dictionary<string, bool>
             {
-                SFProjectRole.CommunityChecker,
-                project.CheckingConfig.CheckingEnabled
-                    && project.CheckingConfig.ShareEnabled
-                    && project.CheckingConfig.ShareLevel == CheckingShareLevel.Anyone
-            },
-            {
-                SFProjectRole.SFObserver,
-                project.TranslateConfig.ShareEnabled && project.TranslateConfig.ShareLevel == TranslateShareLevel.Anyone
-            },
-            {
-                SFProjectRole.Reviewer,
-                project.TranslateConfig.ShareEnabled && project.TranslateConfig.ShareLevel == TranslateShareLevel.Anyone
+                {
+                    SFProjectRole.CommunityChecker,
+                    project.CheckingConfig.CheckingEnabled && project.CheckingConfig.ShareEnabled
+                },
+                { SFProjectRole.SFObserver, project.TranslateConfig.ShareEnabled },
+                { SFProjectRole.Reviewer, project.TranslateConfig.ShareEnabled }
```

Comment:
If you want, you can add a comment to the end of this line, to make any future PRs which add additional roles cleaner and their comparisons simpler.